make sharding strategy configurable and support zero2 algorithm

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74263
* __->__ #73819

adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

zhaojuanmao committed:
[WIP] make sharding strategy configurable and support zero2 algorithm

adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -543,6 +587,9 @@ def _reset_lazy_init(self) -> None:
                 # part of _lazy_init
                 del p._local_shard  # type: ignore[attr-defined]
+        # reset 'reshard_after_forward' flag as well, as it could be changed during
+        # training loop, e.g, it was set as False for root module in _lazy_init().
```
should we support changing this during the training loop? 

Also, if it is set as `False` for root module during _lazy_init but this runs contradictory to the sharding strategy specified by the user, should we actually have that be respected by the root module in `_lazy_init`? Might be confusing to users otherwise.

zhaojuanmao replied:
if users set reshard_after_forward = False, setting it as False for root module does not matter. if users set reshard_after_forward=True, internally setting it as True for root module is kind of internal optimization. Also reshard_after_forward is transparent to users, right? users do not need to know this optimization

rohan-varma replied:
>  internally setting it as True for root module is kind of internal optimization. 

Sorry I might be a bit confused, but do you mean "internally setting it as False"? I believe we currently set reshard_after_forward = False for root module right? 

Maybe a comment explaining this might be useful?

zhaojuanmao replied:
sorry internally setting it as False, refactored comments

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -47,6 +47,35 @@
+                parameters, also inserts reduce_scatter after backward computation for
+                synchronizing and sharding gradients. Sharded optimizer states are
+                updated locally.
+    SHARD_GRAD_OP: Shard optimizer states and gradients, this algorithm inserts all_gather
```
I'm not sure what is the zero-2 implementation in deepspeed. but since we are not free full params after forward, each process could hold the whole model full parameters at the end of forward pass, so it is kind of not sharding parameters here

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -47,6 +47,35 @@
+                   GPU memory until backward computation is done. It inserts reduce_scater
+                   after backward computation for synchronizing and sharding gradients.
+                   Sharded optimizer states are updated locally.
+    NO_SHARD(future support): Parameters, gradients and optimizer states are replicated
```
sounds good

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -47,6 +47,35 @@
+class ShardingStrategy(Enum):
+    """
+    Specify which sharding strategy will be used for the distributed training.
+    FULL_SHARD: Shard parameters, gradients and optimizer states, this algorithm
```
sounds good

zhaojuanmao committed:
Update on "[WIP] make sharding strategy configurable and support zero2 algorithm"


adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_core.py:
```diff
@@ -41,6 +42,23 @@
+    str(ShardingStrategy.SHARD_GRAD_OP): "shard_grad_op",
+}
+
+def subtest_name(*args):
```
Nice! I think in follow-up PRs, we can possibly move these lines to `common_fsdp.py` and use this function in the other `test_fsdp_<...>.py` files to unify the approach to having parsable test names.

I can assign this for myself to do later.

zhaojuanmao replied:
sounds good, thank you

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm.py:
```diff
@@ -75,9 +75,11 @@ def test_communication(
+        #   all-gathers.
         #   Backward: `num_fsdp` - 1 all-gathers (only excluding the root)
-        expected_num_all_gather_no_sync = num_fsdp + (num_fsdp - 1)
+        expected_num_all_gather_no_sync = (num_fsdp - 1) + (num_fsdp - 1)
```
If this is the case, do we understand how this test was passing earlier?

rohan-varma commented on test/distributed/fsdp/test_fsdp_core.py:
```diff
@@ -41,6 +42,23 @@
+}
+
+def subtest_name(*args):
+    return '_'.join([test_name_mapping[str(s)] if s is not None else "none" for s in args])
```
nice approach, let's generalize this and make it more generally useful for tests across distributed as follow up work?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -339,6 +381,19 @@ def __init__(
+            self.reshard_after_forward = False
+        else:
+            raise RuntimeError(
+                "sharding_strategy only supports FULL_SHARD and SHARD_GRAD_OP right now."
```
nit: If we're raising an error here then we might not need to leave things like "# SHARD_OP = auto()" commented out since the user will get this unsupported error. But I think it's fine either way

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1325,7 +1383,15 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+            # self.reshard_after_forward = False (sharding_strategy=SHARD_GRAD_OP),
+            # since the params will not get updated before the next forward,
+            # there is no need to call all_gather to sync params among ranks. This
+            # saves networking bandwidth but uses more GPU memory.
```
A bit confused on this comment - 

regardless of the setting of reshard_after_forward, the param will not get updated before the next forward when in a no_sync manager right? Could we clarify that? 

Also, are we sure that we want this as the default, i.e. the GPU/networking bandwidth trade off is the correct one to make? In general I think we may want to carefully track these tradeoffs that we hardcode in FSDP and eventually expose knobs to allow users to tune (or run experiments and tune / offer recommendations ourselves). What do you think?

zhaojuanmao replied:
sounds good, let me clarify the comments. "reshard_after_forward" is kind of tune knob here, if it is full_shard, no_sync model will do more network communication but save some memory, if it is grad_op, no_sync model will use more memory but save network overhead. let me clarify that in the comment as well

rohan-varma reviewed (approved):
Looks great, excited to have users try out Zero-2! 

Stamping to unblock, but some tests seem to be failing.

zhaojuanmao committed:
Update on "make sharding strategy configurable and support zero2 algorithm"


adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make sharding strategy configurable and support zero2 algorithm"


adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "make sharding strategy configurable and support zero2 algorithm"


adding a new sharding_strategy config in FSDP API to support different data parallel algorithm. also add support for zero2 algorithm, which will only shard optimizer states and grads

Differential Revision: [D34662583](https://our.internmc.facebook.com/intern/diff/D34662583/)

[ghstack-poisoned]
